assets/javascripts/discourse/components/emoji-value-list.js.es6

```diff
@@ -0,0 +1,223 @@
+    later(() => {
+      this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    }, 100);
+  },
+
+  @action
+  clearInput() {
+    this.set("emojiName", "");
+  },
+
+  @on("didReceiveAttrs")
+  _setupCollection() {
+    let object;
+    const values = this.values;
+    const defaultValue = values
+      .split("|")
+      .find(
+        element => element === this.siteSettings.discourse_reactions_like_icon
+      );
+
+    if (!defaultValue) {
+      object = {
+        value: this.siteSettings.discourse_reactions_like_icon,
+        emojiUrl: emojiUrlFor(this.siteSettings.discourse_reactions_like_icon),
+        isLast: false,
+        isEditable: false,
+        isEditing: false
+      };
+    }
+
+    const collectionValues = this._splitValues(values);
+
+    if (object) {
+      collectionValues.unshift(object);
+    }
+
+    this.set("collection", collectionValues);
+  },
+
+  _splitValues(values) {
+    if (values && values.length) {
+      const keys = ["value", "emojiUrl", "isLast"];
+      const res = [];
+      const emojis = values.split("|");
+      emojis.forEach((str, index) => {
+        const object = {};
+        object.value = str;
+
+        if (str === this.siteSettings.discourse_reactions_like_icon) {
+          object.emojiUrl = emojiUrlFor(
+            this.siteSettings.discourse_reactions_like_icon
+          );
+          object.isEditable = false;
+        } else {
+          object.emojiUrl = emojiUrlFor(str);
+          object.isEditable = true;
+        }
+
+        object.isEditing = false;
+
+        object.isLast = emojis.length - 1 === index;
+
+        res.push(object);
+      });
+
+      return res;
+    } else {
+      return [];
+    }
+  },
+
+  @action
+  editValue(index) {
+    const item = this.collection[index];
+    if (item.isEditable) {
+      set(item, "isEditing", !item.isEditing);
+      later(() => {
+        const textbox = document.querySelector(
+          `[data-index="${index}"] .value-input`
+        );
+        if (textbox) {
+          textbox.focus();
+        }
+      }, 100);
+    }
+  },
+
+  actions: {
+    changeValue(index, newValue) {
+      const item = this.collection[index];
+
+      if (this._checkInvalidInput(newValue)) {
+        const oldValues = this.setting.value.split("|");
+
+        if (
+          oldValues.includes(this.siteSettings.discourse_reactions_like_icon)
+        ) {
+          set(item, "value", oldValues[index]);
+        } else {
+          set(item, "value", oldValues[index - 1]);
+        }
+        set(item, "isEditing", !item.isEditing);
+
+        return;
+      }
+
+      this._replaceValue(index, newValue);
+
+      set(item, "isEditing", !item.isEditing);
+    },
+
+    addValue() {
+      if (this._checkInvalidInput([this.emojiName])) {
+        return;
+      }
+      this._addValue(this.emojiName);
+      this.set("emojiName", "");
+    },
+
+    removeValue(value) {
+      this._removeValue(value);
+    },
+
+    shiftUp(index) {
```

Comment:
you could probably have a shiftDirection fn which could be used in shiftUp and shiftDown with just different args

Reply:
I think we can keep both methods separate, as that would be a mess, to check the different conditions required in both methods and combining it into single method.